cloudinit/url_helper.py

```diff
@@ -362,8 +377,79 @@ def _get_ssl_args(url, ssl_details):
     return ssl_args
 
 
+def _get_retry_after(retry_after: str) -> float:
+    """Parse a Retry-After header value into an integer.
+
+    : param retry_after: The value of the Retry-After header.
+        https://www.rfc-editor.org/rfc/rfc9110.html#section-10.2.3
+        https://www.rfc-editor.org/rfc/rfc2616#section-3.3
+    : return: The number of seconds to wait before retrying the request.
+    """
+    try:
+        to_wait = float(retry_after)
+    except ValueError:
+        # Translate a date such as "Fri, 31 Dec 1999 23:59:59 GMT"
+        # into seconds to wait
+        try:
+            to_wait = float(
+                time.mktime(
+                    time.strptime(retry_after, "%a, %d %b %Y %H:%M:%S %Z")
+                )
+                - time.time()
+            )
+        except ValueError:
+            LOG.info(
+                "Failed to parse Retry-After header value: %s. "
+                "Waiting 1 second instead.",
+                retry_after,
+            )
+            to_wait = 1
+        if to_wait < 0:
+            LOG.info(
+                "Retry-After header value is in the past. "
+                "Waiting 1 second instead."
+            )
+            to_wait = 1
+    return to_wait
+
+
+def _handle_error(
```

Comment:
I'm hoping this wouldn't be hard to extend to also modify future request arguments if ever needed in the future...but...YAGNI